### DISCUSSION TITLE
`exec`/`spawn` gets killed but the command still running in background

### DISCUSSION THREAD

**Discussion by yahu1031** [nodejs/help#5008] [Q&A] [answered]
### Version

v19.1.0

### Platform

Darwin Mrs-MacBook-Air.local 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103 arm64

### Subsystem

child_process

### What steps will reproduce the bug?

```js
import { exec, spawn} from 'child_process';
import { join } from 'path';

var execution = exec('ping', { maxBuffer: 1024 * 1024 * 10 });

var spawning = spawn('ping');

// var spawning = spawn('dart', [join('bin', 'main.dart')]);


execution.kill();
```

In my case, I'm running a server. This shows it killed the execution but I can still see the server is up after I killed it. And it also shows that a process in services manager(generic).

### How often does it reproduce? Is there a required condition?

every time.

### What is the expected behavior?

must kill the process.

### What do you see instead?

manually killing it.

### Additional information

NOTE: This is in VS code extension development.

**Comment by bnoordhuis** [comment]
Not a bug so converted to a discussion. Pass `{ killSignal: "SIGKILL" }`. The default signal is `SIGTERM` and a process is free to ignore that.

**Reply by yahu1031** [reply]
@bnoordhuis That doesn't work as expected.

Before killing it.

<img width="561" alt="Screenshot 2022-11-25 at 4 22 54 PM" src="https://user-images.githubusercontent.com/35523357/203967880-ef104e6e-ec77-4ea8-9c1f-e8f5452c2042.png">

After using this code block

```js
process.kill(this._generateSpawn.pid, 'SIGKILL');
this._generateSpawn.kill('SIGKILL');
```

<img width="562" alt="Screenshot 2022-11-25 at 4 24 16 PM" src="https://user-images.githubusercontent.com/35523357/203968173-74511f1f-91fd-4e41-aa34-ed25ba426bdf.png">

It removes the parent process but this sounds like isolating some where

**Reply by yahu1031** [reply]
This command will be watching all my changes and generate a code block. I can now stop my server but not this with a similar code block.

**Reply by bnoordhuis** [reply]
It's not super clear from your example but I suspect you're using `exec()`, not `spawn()`? The former spawns a shell (which then spawns the process), so you're killing the shell, not the grandchild.

**Reply by yahu1031** [reply]
ohhh okie, Lemme cross-check and update you. Thanks for the super quick help.

**Reply by yahu1031** [reply]
It doesn't work still. @bnoordhuis

**Reply by yahu1031** [reply]
<img width="1311" alt="Screenshot 2022-11-25 at 5 08 30 PM" src="https://user-images.githubusercontent.com/35523357/203977733-77ff9f31-77e7-4a31-807f-3804a33d1faa.png">

See the PIDs

**Reply by yahu1031** [reply]
using `spawn()`

**Reply by yahu1031** [reply]
killing the parent doesn't kill the grandchild process but leaves it idle.

**Reply by yahu1031** [reply, marked as answer]
Fixed by using PID ranges `-` in `process.kill(-PID)`
